### ISSUE TITLE
Duplicate translation and filtering holidays

### ISSUE THREAD

**Issue by bisgardo** [vacanza/holidays#2129]
I wanted to add this as a question, but I'm not seeing any button in the discussions pane for adding a post (github also reports an error "You can't perform that action at this time."). But I do think there may be a minor bug involved, so here goes :)

# Context

I'm trying to resolve a subset of holidays of a given country. For example, consider that I want the 2024 holidays of Austria except Easter Monday (2024-04-01).

The way I thought of doing it is by filtering on the name of the holiday. I'm open to suggestions on other strategies.

As the holiday names are being translated by `holidays` to the locale language, the names to ignore have to be translated as well. So the example can be implemented like so:

```python
h = holidays.Austria(years=[2024])

def tr(name):
    return ...

easter_monday = "Ostermontag"
print([ date for date, name in h.items() if name is not tr(easter_monday) ])
```
where `tr` is the appropriate function to align with the translations done by `holidays`.

Due to duplicate translation (explained below), it seems to me like I have to translate the name using both `gettext` and `h.tr` in order to be certain that I get the exact same transformation:

```python
from gettext import gettext

def tr(name):
    return gettext(h.tr(name))
```

# Duplicate translation

When the holidays of a given country is initialized, the holiday names are translated using `gettext`. For example, for [Austria](https://github.com/vacanza/holidays/blob/d510248866534b78a4c79f1d06a615aa44b4ad37/holidays/countries/austria.py#L71), new year's day is added using

```python
self._add_new_years_day(tr("Neujahr"))
```
where `tr` is an alias of `gettext`.

However, this call ends up in [`_add_holiday` of `HolidayBase`](https://github.com/vacanza/holidays/blob/d510248866534b78a4c79f1d06a615aa44b4ad37/holidays/holiday_base.py#L749) where it does
```python
self[dt] = self.tr(name)
```
where [`self.tr`](https://github.com/vacanza/holidays/blob/d510248866534b78a4c79f1d06a615aa44b4ad37/holidays/holiday_base.py#L364) is also defined as `gettext` (though configured to use the translations shipped with the library). That is, the name is translated twice, first before being passed to `_add_new_years_day` (using "plain" `gettext`), and then again in a nested call (using the library translations).

# Question

Do I understand correctly that the string is being translated twice? If so, why is that? The first call to `gettext` looks redundant to me. Or if it isn't, it could be done from within `self.tr`.

Also, am I solving the original problem in a reasonable way?

**Comment by bisgardo** [comment]
@arkid15r Thank you so much for the detailed answer.

> You might be looking for this link: [Start a new discussion in Q&A](https://github.com/vacanza/holidays/discussions/new?category=q-a)

This is how the page looks to me. The button for creating a post should be right next to the filter one, right?

![image](https://github.com/user-attachments/assets/d09738f3-6dd3-43b8-917e-b3426ce3c95e)

> To filter holidays, you can use either **date** or **holiday name**. Filtering by date is straightforward -- you can simply remove the key from the dict-like object. For filtering by name, consider using the auxiliary method [`pop_named`](https://holidays.readthedocs.io/en/latest/api.html#holidays.holiday_base.HolidayBase.pop_named). Here's an example:
> 
> ```python
> h = holidays.Austria(years=2024, language="de")
> h.pop_named("Ostermontag")
> ```

Based on this, I'm now using this implementation which I like much better:

```python

def holidays_by_region(region: str, year_start, year_end, names_to_ignore: List[str]):
    # Resolve regional class.
    h = getattr(holidays, region)
    # Instantiate class for holding the holidays while
    # keeping the names in original language.
    # This avoids locale-dependent behavior when filtering below.
    dates = h(years=range(year_start, year_end), language=h.default_language)
    # Exclude ignored
    for n in names_to_ignore:
        # Look up date by name.
        # Not using 'pop_named' because that does lookup using partial match.
        for d in dates.get_named(n, lookup="exact", split_multiple_names=False):
            dates.pop(d)
    return dates.keys()
```

> You should use `!=` instead of `is not`. Here's why: [Python: != vs is not](https://stackoverflow.com/questions/2209755/python-operation-vs-is-not).

Thanks for pointing this out - I'm not yet entirely native in Python ;)

> As for the duplicate translation question -- it’s a bit more nuanced. The actual translation is handled by `self.tr`, while `tr` acts as a sort of marker (I [implemented this](https://github.com/vacanza/holidays/pull/827) some time ago).

As I see it, the actual translation happens in `self.tr` as you say. But can you please elaborate on what you mean by `tr` being a "marker"? It's [an alias of `gettext`](https://github.com/vacanza/holidays/blob/d510248866534b78a4c79f1d06a615aa44b4ad37/holidays/countries/austria.py#L13), so it will use system-level translations if it finds them, right?

> For better code readability, I removed the `self.` prefix wherever possible. And we'd gladly consider any PRs with improvement suggestions if you have ideas!

I'm also a little confused by this: There's nowhere where `tr` and `self.tr` refers to the same thing?

> I hope I addressed all your questions and concerns. If there's anything you'd like me to clarify further, please don't hesitate to let me know. Apologies for the delayed response -- I’ve been a bit busy recently with KubeCon'24.

You don't owe me anything, so no need for apologies!

**Comment by arkid15r** [comment]
>This is how the page looks to me. The button for creating a post should be right next to the filter one, right?

I'm not sure, try clicking on Q&A

> \# Not using 'pop_named' because that does lookup using partial match.

Sounds like a feature request to me.

> But can you please elaborate on what you mean by `tr` being a "marker"?

I had to look this up actually :) I meant this [particular line of code](https://github.com/vacanza/holidays/blob/88f5c65590c0521a3f1d38213678578191406589/scripts/l10n/generate_po_files.py#L76) -- we adopted [lingua](https://github.com/vacanza/lingva) after it had been abandoned.

>You don't owe me anything, so no need for apologies!

👍 I know it's open source and all. After taking over this project the core team is the best point of contact for clarifying `holidays` related questions. I try my best to cover issues as a last resort point. I hope I explained it right.
